ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Peras/Voting/Rules.hs

```diff
@@ -0,0 +1,398 @@
+    VR2B ->
+      "the last certificate on chain is exactly one or more cooldown periods old"
+
+-- | VR-1A: the voter has seen the certificate for the previous round, and the
+-- certificate was received in the first X slots after the start of the round.
+perasVR1A ::
+  HasPerasCertRound cert =>
+  PerasVotingView cert ->
+  Pred PerasVotingRule
+perasVR1A
+  PerasVotingView
+    { perasParams
+    , currRoundNo
+    , latestCertSeen
+    , arrivalSlot
+    , certRoundStart
+    } =
+    VR1A := vr1a1 :/\: vr1a2
+   where
+    -- The latest certificate seen is from the previous round
+    --
+    -- NOTE: 'succWithOrigin' handles the 'Origin' case (i.e. when we have
+    -- never seen a certificate before) correctly by returning 0, as this is
+    -- the only round number that should satisfy this equality when we are
+    -- bootstrapping the voting process. In other words, we should be able to
+    -- start voting from round 0 even if we have never seen a certificate
+    -- before, but failing to do so should trigger a cooldown period
+    -- immediately after.
+    vr1a1 =
+      currRoundNo :==: succWithOrigin (getPerasCertRound <$> latestCertSeen)
+
+    -- The latest certificate seen was received within X slots from the start
+    -- of the round
+    vr1a2 =
+      case latestCertSeen of
+        -- We have seen a certificate ==> check its arrival time
+        NotOrigin cert -> arrivalSlot cert :<=: certRoundStart cert + _X
+        -- We have never seen a certificate ==> vacuously true
+        Origin -> Bool True
+
+    _X =
+      SlotNo $
+        unPerasCertArrivalThreshold $
+          perasCertArrivalThreshold perasParams
+
+-- | VR-1B: the block being voted upon extends the most recently certified one.
+perasVR1B ::
+  PerasVotingView cert ->
+  Pred PerasVotingRule
+perasVR1B
+  PerasVotingView
+    { latestCertSeen
+    , candidateExtendsCert
+    } =
+    VR1B := vr1b
+   where
+    -- The block being voted upon extends the most recently certified one
+    vr1b =
+      case latestCertSeen of
+        -- We have seen a certificate ==> check that it extends our chain
+        NotOrigin cert -> Bool (candidateExtendsCert cert)
+        -- We have never seen a certificate ==> vacuously true
+        Origin -> Bool True
+
+-- | VR-2A: the last certificate a party has seen is from a round at least R
+-- rounds previously. This enforces the chain-healing period that must occur
+-- before leaving a cool-down period.
+perasVR2A ::
+  HasPerasCertRound cert =>
+  PerasVotingView cert ->
+  Pred PerasVotingRule
+perasVR2A
+  PerasVotingView
+    { perasParams
+    , currRoundNo
+    , latestCertSeen
+    } =
+    VR2A := vr2a
+   where
+    vr2a =
+      -- NOTE: we use 'succWithOrigin' and '-1' to handle the 'Origin' case
+      -- (i.e. when we have never seen a certificate before) correctly,
+      -- treating the 'Origin' certificate as being from round -1.
+      (succWithOrigin (getPerasCertRound <$> latestCertSeen) - 1 + _R)
```

Comment:
Why don't we use a `Origin -> Bool True` case here as we did previously? IMO it would make the code clearer.

Reply:
Turned all 'succWithOrigin` into explicit `case` expressions.